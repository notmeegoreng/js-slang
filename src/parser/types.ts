import type { Program } from 'estree'

import type { Context, Node, Chapter, Variant, SourceError } from '../types'

export type { ParserOptions as BabelOptions } from '@babel/parser'
export type { Options as AcornOptions } from 'acorn'

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}
export interface Rule<T extends Node> {
  /**
   * Name of the rule
   */
  name: string

  /**
   * Test snippets to test the behaviour of the rule. Providing no test snippets
   * means that this rule will not be tested when running unit tests.\
   * First element of the tuple is the code to test. Set the second element to `undefined`
   * if the snippet should not throw an error. Otherwise set it to the `explain()` value
   * of the error.
   */
  testSnippets?: [code: string, expected: string | undefined][]

  /**
   * Disable this rule for this chapter (inclusive) and above
   */
  disableFromChapter?: Chapter
  disableForVariants?: Variant[]
  checkers: {
    [name: string]: (node: T, ancestors: Node[]) => SourceError[]
  }
}
