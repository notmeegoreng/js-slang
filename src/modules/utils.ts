import * as _ from 'lodash'

import { RecursivePartial } from '../types'
import { ImportOptions } from './moduleTypes'
import { defaultAnalysisOptions } from './preprocessor/analyzer'

const exportDefaultStr = 'export default'
export function removeExportDefault(text: string) {
  if (text.startsWith(exportDefaultStr)) {
    text = text.substring(exportDefaultStr.length).trim()
  }

  return text
}

export function mergeImportOptions(src?: RecursivePartial<ImportOptions>): ImportOptions {
  const baseOptions = _.cloneDeep(defaultAnalysisOptions)
  return _.merge(baseOptions, src as any)
}

export const isSourceModule = (path: string) => !path.startsWith('.') && !path.startsWith('/')
