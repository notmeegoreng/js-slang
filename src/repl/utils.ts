import { Option } from '@commander-js/extra-typings'
import { Chapter, Variant } from '../types'

export function chapterParser(str: string): Chapter {
  let foundChapter: string | undefined

  if (/^[0-9]$/.test(str)) {
    // Chapter is fully numeric
    const value = parseInt(str)
    foundChapter = Object.keys(Chapter).find(chapterName => Chapter[chapterName] === value)
  } else {
    foundChapter = str
  }

  if (foundChapter === undefined) {
    throw new Error(`Invalid chapter value: ${str}`)
  }

  return Chapter[foundChapter]
}

export const chapterOption = new Option('--chapter <chapter>')
  .default(Chapter.SOURCE_4)
  .argParser(chapterParser)

export const variantOption = new Option('--variant <variant>')
  .default(Variant.DEFAULT)
  .choices(Object.values(Variant))
