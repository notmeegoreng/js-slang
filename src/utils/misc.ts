import { RuntimeSourceError } from '../errors/runtimeSourceError'

export class PromiseTimeoutError extends RuntimeSourceError {}

export const timeoutPromise = <T>(promise: Promise<T>, timeout: number) =>
  new Promise<T>((resolve, reject) => {
    const timeoutid = setTimeout(() => reject(new PromiseTimeoutError()), timeout)

    promise
      .then(res => {
        clearTimeout(timeoutid)
        resolve(res)
      })
      .catch(e => {
        clearTimeout(timeoutid)
        reject(e)
      })
  })

/**
 * Run the mapping function over the items, filtering out any items that
 * that the mapping function returned `undefined` for
 */
export function mapAndFilter<T, U>(items: T[], mapper: (input: T) => U | undefined) {
  return items.reduce((res, item) => {
    const newItem = mapper(item)
    if (newItem !== undefined) return [...res, newItem]
    return res
  }, [] as U[])
}

export function mapObject<T extends Record<any, any>, U>(
  obj: T,
  mapper: (key: keyof T, value: T[keyof T]) => U
): Record<keyof T, U> {
  return objectEntries(obj).reduce(
    (res, [key, value]) => ({
      ...res,
      [key]: mapper(key, value)
    }),
    {} as Record<keyof T, U>
  )
}

/**
 * Type safe `Object.keys`
 */
export function objectKeys<T extends string | number | symbol>(obj: Record<T, any>): T[] {
  return Object.keys(obj) as T[]
}

/**
 * Type safe `Object.values`
 */
export function objectValues<T>(obj: Record<any, T>) {
  return Object.values(obj) as T[]
}

type DeconstructRecord<T extends Record<any, any>> = {
  [K in keyof T]: [K, T[K]]
}

type EntriesOfRecord<T extends Record<any, any>> = DeconstructRecord<T>[keyof T]

export function objectEntries<T extends Record<any, any>>(obj: T): EntriesOfRecord<T>[] {
  return Object.entries(obj)
}
