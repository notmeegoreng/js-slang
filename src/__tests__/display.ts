import { runInContext } from '..'
import { Chapter } from '../types'
import { stripIndent } from '../utils/formatters'
import { createTestContext, expectFinishedResult, expectParsedError } from '../utils/testing'
import { testMultipleCases } from '../utils/testing/testers'

testMultipleCases<[string, ...string[]]>(
  [
    [
      'display second argument can be a string',
      `display(31072020, "my_first_String");`,
      'my_first_String 31072020'
    ],
    ['display can be used to display numbers', 'display(0);', '0'],
    [
      'display can be used to display funny numbers',
      'display(1e38); display(NaN); display(Infinity);',
      '1e+38',
      'NaN',
      'Infinity'
    ],
    [
      'display can be used to display (escaped) strings',
      `display("Tom's assistant said: \\"tuna.\\"");`,
      '"Tom\'s assistant said: \\"tuna.\\""'
    ],
    [
      'raw_display can be used to displayed (unescaped) strings directly',
      `raw_display("Tom's assisstant said: \\"tuna.\\"");`,
      'Tom\'s assisstant said: "tuna."'
    ],
    [
      'display can be used to display arrow functions',
      `display(x => x); display((x, y) => x + y);`,
      'x => x',
      '(x, y) => x + y'
    ],
    [
      'display can be used to display functions',
      `
      function f() { return 0; }
      display(f);
    `,
      stripIndent`
      function f() {
        return 0; 
      }
    `
    ],
    [
      'displaying builtins hides their implementation',
      'display(pair);',
      stripIndent`
      function pair(left, right) {
        [implementation hidden]
      }
    `
    ],
    ['display can be used with lists', 'display(list(1, 2));', '[1, [2, null]]'],
    ['display can be used with arrays', 'display([1, 2, [4, 5]]);', '[1, 2, [4, 5]]'],
    [
      'display can be used with objects',
      `display({a: 1, b: 2, c: {d: 3}});`,
      '{"a": 1, "b": 2, "c": {"d": 3}}'
    ]
  ],
  async ([code, ...expected]) => {
    const context = createTestContext({ chapter: Chapter.LIBRARY_PARSER })
    const result = await runInContext(code, context)

    expectFinishedResult(result)
    expect(context.displayResult).toEqual(expected)
  }
)

test('display throw error if second argument is non-string when used', () => {
  return expectParsedError(`display(31072020, 0xDEADC0DE);`).toMatchInlineSnapshot(
    `"Line 1: TypeError: display expects the second argument to be a string"`
  )
})

test('display with no arguments throws an error', () => {
  return expectParsedError(`display();`, { chapter: Chapter.LIBRARY_PARSER }).toMatchInlineSnapshot(
    `"Line 1: Expected 1 or more arguments, but got 0."`
  )
})
