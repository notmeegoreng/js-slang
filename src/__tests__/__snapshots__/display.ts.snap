// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display throw error if second argument is non-string when used: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "display(31072020, 0xDEADC0DE);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 1: TypeError: display expects the second argument to be a string",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`display with no arguments throws an error: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "display();",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 1: Expected 1 or more arguments, but got 0.",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;
